fem/linearform_ext.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "linearform.hpp"
+#include "../general/forall.hpp"
+
+// Implementations of FullLinearFormExtension.
+
+namespace mfem
+{
+
+FullLinearFormExtension::FullLinearFormExtension(LinearForm *lf):
+   LinearFormExtension(lf)
+{
+   Update();
+}
+
+void FullLinearFormExtension::Assemble()
+{
+   MFEM_VERIFY(lf->Size() == lf->FESpace()->GetVSize(), "");
+
+   // Filter out the unsupported integrators
+   MFEM_VERIFY(lf->GetBLFI()->Size() == 0,
+               "Integrators added with AddBoundaryIntegrator() "
+               "are not supported!");
+
+   MFEM_VERIFY(lf->GetDLFI_Delta()->Size() == 0, ""
+               "Integrators added with AddDomainIntegrator() which are "
+               "DeltaLFIntegrators with delta coefficients "
+               "are not supported!");
+
+   MFEM_VERIFY(lf->GetIFLFI()->Size() == 0,
+               "Integrators added with AddInteriorFaceIntegrator() "
+               "are not supported!");
+
+   MFEM_VERIFY(lf->GetFLFI()->Size() == 0,
+               "Integrators added with AddBdrFaceIntegrator() "
+               " are not supported!");
+
+   const FiniteElementSpace &fes = *lf->FESpace();
+   const Array<Array<int>*> &domain_integs_marker = *lf->GetDLFIM();
+   const int mesh_attributes_size = fes.GetMesh()->attributes.Size();
+   const Array<LinearFormIntegrator*> &domain_integs = *lf->GetDLFI();
+
+   for (int k = 0; k < domain_integs.Size(); ++k)
+   {
+      // Get the markers for this integrator
+      const Array<int> *domain_integs_marker_k = domain_integs_marker[k];
+
+      // check if there are markers for this integrator
+      const bool has_markers_k = domain_integs_marker_k != nullptr;
+
+      if (has_markers_k)
+      {
+         // Element attribute marker should be of length mesh->attributes
+         MFEM_VERIFY(mesh_attributes_size == domain_integs_marker_k->Size(),
+                     "invalid element marker for domain linear form "
+                     "integrator #" << k << ", counting from zero");
+      }
+
+      const int NE = fes.GetNE();
+      auto markers_w = markers.Write();
+
+      // if there are no markers, just use the whole linear form (1)
+      if (!has_markers_k)
+      {
+         // done this way as operator= for array are still done on the host
```

Comment:
Can you elaborate? I added the `UseDevice(true)` on the `markers`, but this portion of the code still needs to be done (`=1`).

Reply:
I would think that with `UseDevice(true)` the `operator=` is done on device, is that not the case?

Reply:
Class `Array` does not implement any operation on device at the moment. The reason for having the method `Array::UseDevice(true)` is to enable device execution in `Vector` kernels like `Vector::GetSubVector`, `Vector::SetSubVector`, etc.

Reply:
I guess we never defined `Array::UseDevice(bool)` (I also forgot about that) -- one can always use `my_array.GetMemory().UseDevice(true)` or just `my_array.Read()` or `my_array.Write()`.

The reason that there are no device-enabled operations like `operator=()` is that we never had a widespread need for them. However, adding such support is easy if there is need.

The use of `Memory` inside the `Array` class is needed in many places -- e.g. any place where we want to use an integer array on device, we call `Array::Read()`, `Array::ReadWrite()`, or `Array::Write()` which are implemented using the `Memory` class member `data` of class `Array`.

Reply:
Then since the need is here now, instead of having the code for `operator=` inlined here, I would be in favor of adding a `UseDevice(bool)` method to the `Array` class, and moving this code in `operator=`.
It also seems that currently the `Array` class is very unsafe as there is no `MFEM_VERIFY` nor `MFEM_ASSERT` to check that methods processing on host are processing up-to-date data.

Reply:
This sounds reasonable to me.

Reply:
Added `void UseDevice(bool use_dev)` in `Array.hpp`

Reply:
I'm not sure about adding the logic in `inline void Array<T>::operator=(const T &a)` in this PR.

Reply:
Isn't it simply:
```c++
template <class T>
inline void Array<T>::operator=(const T &a)
{
   const bool use_dev = UseDevice();
   const int N = size;
   auto y = Write(data, size, on_dev);
   const T value = a;
   MFEM_FORALL_SWITCH(use_dev, i, N, y[i] = value;);
}
```
?
Are you worried that it may break some code?

I'm okay to do this in another PR that fixes the unsafe behavior of `Array<T>`, since it would probably be better to do this for all methods that can run on device, and to `MFEM_ASSERT` synchronization (or `Read()`) for methods that currently assume that the data is always valid on host.

Reply:
`done`

Reply:
What is done?